ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGas.java

```diff
@@ -51,21 +60,94 @@ public String getName() {
 
   @Override
   public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+
     final JsonCallParameter callParams =
         requestContext.getRequiredParameter(0, JsonCallParameter.class);
 
     final BlockHeader blockHeader = blockHeader();
     if (blockHeader == null) {
-      return errorResponse(requestContext, JsonRpcError.INTERNAL_ERROR);
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INTERNAL_ERROR);
     }
+    return doEstimateGas(requestContext.getRequest(), callParams, blockHeader);
+  }
+
+  /**
+   * Allows to estimate gas for a transaction
+   *
+   * @param request json rpc request
+   * @param callParams call params
+   * @param blockHeader block header
+   * @return json rpc response with the estimate gas of the transaction, otherwise the error
+   */
+  private JsonRpcResponse doEstimateGas(
+      final JsonRpcRequest request,
+      final JsonCallParameter callParams,
+      final BlockHeader blockHeader) {
 
-    final JsonCallParameter modifiedCallParams =
+    long lowGasLimit, midGasLimit, highGasLimit;
+    int numberOfRetry = 0;
+    boolean foundEstimateGas;
+
+    JsonCallParameter modifiedCallParams =
         overrideGasLimitAndPrice(callParams, blockHeader.getGasLimit());
 
-    return transactionSimulator
-        .process(modifiedCallParams, blockHeader.getNumber())
-        .map(gasEstimateResponse(requestContext))
-        .orElse(errorResponse(requestContext, JsonRpcError.INTERNAL_ERROR));
+    Optional<TransactionSimulatorResult> simulatorResult = Optional.empty();
+
+    try {
+
+      // make a first estimate of the necessary gasLimit
+      simulatorResult = transactionSimulator.process(modifiedCallParams, blockHeader.getNumber());
+      lowGasLimit =
+          highGasLimit =
+              simulatorResult
+                  .filter(TransactionSimulatorResult::isSuccessful)
+                  .map(TransactionSimulatorResult::getGasEstimate)
+                  .orElseThrow();
+
+      // check that the estimate is valid and if not we increase the estimate by
+      // GAS_ESTIMATE_CHANGE_DENOMINATOR. if after MAX_ESTIMATE_NUMBER_OF_RETRY retries you cannot
+      // find a valid estimate. we return the error returned by the transaction
+      do {
+
+        modifiedCallParams.setGasLimit(highGasLimit);
```

Comment:
I'd prefer we clone and change the object instead of introducing a single mutable field into an otherwise immutable class.

```suggestion
JsonCallParameter modifiedCallParams =
        overrideGasLimitAndPrice(callParams, highGasLimit);
```